### PR TITLE
ACTIN-96: Sort signatures by allocation and add etiology

### PR DESCRIPTION
**Pull request by cbruel**

### CODE REVIEW

File: hmf-common/src/test/java/com/hartwig/hmftools/common/sigs/SignatureTestFactory.java
Comment (kduyvesteyn) [inline hmf-common/src/test/java/com/hartwig/hmftools/common/sigs/SignatureTestFactory.java:1]: I would not adjust this test factory, this relates to the hmf-common datamodel (sigs) and not the orange datamodel.

File: orange-datamodel/src/main/java/com/hartwig/hmftools/datamodel/sigs/SignatureAllocation.java [hunk 1]
```diff
@@ -16,4 +16,7 @@ public interface SignatureAllocation
     double allocation();
 
     double percent();
+
+    @Nullable
```
Comment (kduyvesteyn) [inline orange-datamodel/src/main/java/com/hartwig/hmftools/datamodel/sigs/SignatureAllocation.java:20]: Why would this ever be null? 

I can imagine there is no etiology for `MISALLOC` but maybe you can assign "-" in the datamodel in that case? Since etiology is not for display purpose anyways I think I would avoid null

File: orange/src/main/java/com/hartwig/hmftools/orange/OrangeConfig.java [hunk 2]
```diff
@@ -73,6 +73,7 @@ public interface OrangeConfig
     String DOID_JSON = "doid_json";
     String COHORT_MAPPING_TSV = "cohort_mapping_tsv";
     String COHORT_PERCENTILES_TSV = "cohort_percentiles_tsv";
+    String SIGNATURES_ETIOLOGIES_TSV = "signatures_etiologies_tsv";
```
Comment (kduyvesteyn) [inline orange/src/main/java/com/hartwig/hmftools/orange/OrangeConfig.java:76]: In general I believe the correct term is "signature_etiologies". 

There is 1 signature_etiology
There are N signature_etiologies 

:)

(affects naming in multiple places)

File: orange/src/main/java/com/hartwig/hmftools/orange/algo/OrangeAlgo.java [hunk 3]
```diff
@@ -106,6 +107,8 @@ public class OrangeAlgo
     @NotNull
     private final List<DriverGene> driverGenes;
     @NotNull
+    private final Map<String, String> signaturesEtiologies;
```
Comment (kduyvesteyn) [inline orange/src/main/java/com/hartwig/hmftools/orange/algo/OrangeAlgo.java:110]: Maybe a more correct name would be "etiologyPerSignature". That makes it more clear your keys are signatures and the values are the etiologies?

File: orange/src/main/java/com/hartwig/hmftools/orange/algo/OrangeAlgo.java [hunk 4]
```diff
@@ -150,18 +157,20 @@ public static OrangeAlgo fromConfig(@NotNull OrangeConfig config) throws IOExcep
         String outputDir = config.outputDir();
         PlotManager plotManager = !outputDir.isEmpty() ? new FileBasedPlotManager(outputDir) : new DummyPlotManager();
 
-        return new OrangeAlgo(doidEntry, mapper, percentilesModel, driverGenes, knownFusionCache, ensemblDataCache, plotManager);
+        return new OrangeAlgo(doidEntry, mapper, percentilesModel, driverGenes, signaturesEtiologies, knownFusionCache, ensemblDataCache, plotManager);
```
Comment (kduyvesteyn) [inline orange/src/main/java/com/hartwig/hmftools/orange/algo/OrangeAlgo.java:160]: Minor but this line extends max length, should be auto-reformatted.

File: orange/src/main/java/com/hartwig/hmftools/orange/algo/sigs/SigsEtiologiesLoader.java [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+import java.io.IOException;
+import java.nio.file.Files;
+import java.util.List;
+import java.util.Map;
+
+import com.google.common.collect.Maps;
+
+import org.jetbrains.annotations.NotNull;
+
+public class SigsEtiologiesLoader
```
Comment (kduyvesteyn) [inline orange/src/main/java/com/hartwig/hmftools/orange/algo/sigs/SigsEtiologiesLoader.java:16]: Class should be final (not meant to be extended, its the equivalent of "object" in kotlin)

File: orange/src/main/java/com/hartwig/hmftools/orange/report/tables/SignatureAllocationTable.java [hunk 6]
```diff
@@ -72,10 +75,7 @@ else if(sig2.equalsIgnoreCase(MISALLOC_SIGNATURE))
                 return sig1.compareTo(sig2);
             }
 
-            int sigNumber1 = Integer.parseInt(sig1.substring(3));
-            int sigNumber2 = Integer.parseInt(sig2.substring(3));
-
-            return Integer.compare(sigNumber1, sigNumber2);
+            return Double.compare(allocation1.allocation(), allocation2.allocation()) * -1;
```
Comment (kduyvesteyn) [inline orange/src/main/java/com/hartwig/hmftools/orange/report/tables/SignatureAllocationTable.java:78]: It is more typical to use `Double.compare(allocation2, allocation1)` if you want to order descending.

File: orange/src/main/java/com/hartwig/hmftools/orange/conversion/OrangeConversion.java [hunk 7]
```diff
@@ -154,13 +159,22 @@ public static PeachGenotype convert(@NotNull com.hartwig.hmftools.common.peach.P
                 .build();
     }
 
-    @NotNull
-    public static SignatureAllocation convert(@NotNull com.hartwig.hmftools.common.sigs.SignatureAllocation signatureAllocation)
+    @Nullable
+    public static List<SignatureAllocation> convertAndAnnotateWithEtiology(
+            @Nullable List<com.hartwig.hmftools.common.sigs.SignatureAllocation> signatureAllocation,
```
Comment (kduyvesteyn) [inline orange/src/main/java/com/hartwig/hmftools/orange/conversion/OrangeConversion.java:164]: Why can signatureAllocation be null?

File: orange/src/main/java/com/hartwig/hmftools/orange/conversion/OrangeConversion.java
Comment (kduyvesteyn) [inline orange/src/main/java/com/hartwig/hmftools/orange/conversion/OrangeConversion.java:1]: I would not put the the etiology annotation in this class. This class is meant for trivial conversion. 

I would suggest to make a new class "SigsInterpreter" that implements this logic

File: orange/src/main/java/com/hartwig/hmftools/orange/report/tables/SignatureAllocationTable.java [hunk 8]
```diff
@@ -33,13 +34,15 @@ public static Table build(@NotNull String title, float width, @NotNull List<Sign
...
-                new Cell[] { cells.createHeader("Signature"), cells.createHeader("Allocation"), cells.createHeader("Percent"),
+                new float[] { 1, 3, 1, 1, 1 },
+                new Cell[] { cells.createHeader("Signature"), cells.createHeader("Etiology"), cells.createHeader("Allocation"),
+                        cells.createHeader("Percent"),
                         cells.createHeader(Strings.EMPTY) });
 
         for(SignatureAllocation signatureAllocation : sort(signatureAllocations))
         {
             table.addCell(cells.createContent(signatureAllocation.signature()));
+            table.addCell(cells.createContent(Objects.requireNonNullElse(signatureAllocation.etiology(), "-")));
```
Comment (kduyvesteyn) [inline orange/src/main/java/com/hartwig/hmftools/orange/report/tables/SignatureAllocationTable.java:45]: As mentioned before, I would set "-" for MISALLOC in the algo and make the etiology itself NotNull

File: orange/src/test/java/com/hartwig/hmftools/orange/report/tables/SignatureAllocationTableTest.java [hunk 9]
```diff
@@ -16,24 +16,24 @@ public class SignatureAllocationTableTest
...
-        assertEquals(allocation2, sorted.get(1));
+        assertEquals(allocation2, sorted.get(0));
+        assertEquals(allocation1, sorted.get(1));
         assertEquals(allocation3, sorted.get(2));
         assertEquals(allocation4, sorted.get(3));
     }
 
     @NotNull
-    private static SignatureAllocation create(@NotNull String signature)
+    private static SignatureAllocation create(@NotNull String signature, @NotNull Double allocation)
```
Comment (kduyvesteyn) [inline orange/src/test/java/com/hartwig/hmftools/orange/report/tables/SignatureAllocationTableTest.java:35]: `double` is the same as `NotNull Double`

File: orange/src/test/resources/sigs/signatures_etiologies.tsv
Comment (kduyvesteyn) [inline orange/src/test/resources/sigs/signatures_etiologies.tsv:1]: I would put the entire file in the test resources.

Imagine in a future pipeline run COLO gets a signature that is not in this file - we would have to update it again.

Also, I would put this file in `test_run_resources` to make it clear it is there to support the `test_run` ORANGE report

File: orange-datamodel/src/main/java/com/hartwig/hmftools/datamodel/sigs/SignatureAllocation.java (hunk 1 above)
Comment (cbruel) [reply orange-datamodel/src/main/java/com/hartwig/hmftools/datamodel/sigs/SignatureAllocation.java:20]: Indeed my reasoning behind making it nullable was because of `MISALLOC`

File: orange/src/main/java/com/hartwig/hmftools/orange/conversion/OrangeConversion.java (hunk 7 above)
Comment (cbruel) [reply orange/src/main/java/com/hartwig/hmftools/orange/conversion/OrangeConversion.java:164]: This input comes from `loadSigAllocations`, which could have a nullable output (this could e.g. happen when no sigs allocation tsv is provided)

Comment (kduyvesteyn) [reply orange/src/main/java/com/hartwig/hmftools/orange/conversion/OrangeConversion.java:164]: Clear! Didn't realise you changed the input from a single sig to a list of sigs, my bad!

File: hmf-common/src/test/java/com/hartwig/hmftools/common/sigs/SignatureTestFactory.java
Comment (cbruel) [reply hmf-common/src/test/java/com/hartwig/hmftools/common/sigs/SignatureTestFactory.java:1]: Reverted the change. But `SignatureTestFactory` is now not used anymore (now it isn't used anymore in `OrangeConversionTest`). So I would say the factory could be removed?

File: orange/src/test/resources/sigs/signatures_etiologies.tsv
Comment (cbruel) [reply orange/src/test/resources/sigs/signatures_etiologies.tsv:1]: I've put the entire file in the test resources, and uploaded the file in `common-resources-public`.

File: orange/src/main/java/com/hartwig/hmftools/orange/algo/sigs/SigsEtiologiesLoader.java (hunk 5 above)
Comment (cbruel) [reply orange/src/main/java/com/hartwig/hmftools/orange/algo/sigs/SigsEtiologiesLoader.java]: I've also added a `SigsEtiologiesLoaderTest` btw, since I thought it was missing

File: orange-datamodel/src/main/java/com/hartwig/hmftools/datamodel/sigs/SignatureAllocation.java [hunk 10]
```diff
@@ -16,4 +16,6 @@ public interface SignatureAllocation
     double allocation();
 
     double percent();
+
+    String etiology();
```
Comment (kduyvesteyn) [inline orange-datamodel/src/main/java/com/hartwig/hmftools/datamodel/sigs/SignatureAllocation.java:20]: Should be NotNull :)

Also, I think I would put this property under "signature" (same as what you have done in the example jsons, feels like natural order.

File: orange/src/main/java/com/hartwig/hmftools/orange/algo/sigs/SigsInterpreter.java [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+            @NotNull Map<String, String> etiologyPerSignature)
+    {
+        if(signatureAllocation == null)
+        {
+            return null;
+        }
+        return signatureAllocation.stream()
+                .map(signature -> ImmutableSignatureAllocation.builder()
+                        .signature(signature.signature())
+                        .etiology(etiologyPerSignature.getOrDefault(signature.signature(), "-"))
```
Comment (kduyvesteyn) [inline orange/src/main/java/com/hartwig/hmftools/orange/algo/sigs/SigsInterpreter.java:27]: I would think this is fine, but you may also want to consider adding some extra safeties. Imagine Sigs changes in the meantime and suddenly writes "sig1" rather than "Sig1". Would you want to get "-" etiology without being aware? 

You could also consider mapping MISALLOC to "-", and raise a warning in case any other signature is missing from the map? 

I don't have a super strong opinion though, could go either way

File: orange/src/main/java/com/hartwig/hmftools/orange/algo/sigs/SigsInterpreter.java [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+import com.hartwig.hmftools.datamodel.sigs.ImmutableSignatureAllocation;
+import com.hartwig.hmftools.datamodel.sigs.SignatureAllocation;
+
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+public final class SigsInterpreter
+{
+    @Nullable
+    public static List<SignatureAllocation> convertAndAnnotateWithEtiology(
```
Comment (kduyvesteyn) [inline orange/src/main/java/com/hartwig/hmftools/orange/algo/sigs/SigsInterpreter.java:16]: I would rename function to `interpret' to be more compatible with e.g. PurpleInterpreter

File: orange/src/main/java/com/hartwig/hmftools/orange/report/tables/SignatureAllocationTable.java [hunk 13]
```diff
@@ -33,13 +33,15 @@ public static Table build(@NotNull String title, float width, @NotNull List<Sign
 
         Cells cells = new Cells(reportResources);
         Table table = Tables.createContent(width,
-                new float[] { 1, 1, 1, 3 },
-                new Cell[] { cells.createHeader("Signature"), cells.createHeader("Allocation"), cells.createHeader("Percent"),
+                new float[] { 1, 2.3F, 1, 1, 1 },
```
Comment (kduyvesteyn) [inline orange/src/main/java/com/hartwig/hmftools/orange/report/tables/SignatureAllocationTable.java:36]: One note about these numbers: They are all ratios, and I typically don't use floating point numbers.

If you multiple every number by 10 you get the same table (10, 23, 10, 10, 10) which I think would be a bit cleaner (but its a matter of taste!)

File: orange/src/test/java/com/hartwig/hmftools/orange/algo/sigs/SigsInterpreterTest.java [hunk 14]
```diff
@@ -0,0 +1,31 @@
...
+
+    static
+    {
+        etiologyPerSignature.put("Sig1", "APOBEC");
+    }
+
+    @Test
+    public void shouldConvertAndAnnotateSignatures()
+    {
+        assertNotNull(SigsInterpreter.convertAndAnnotateWithEtiology(signature, etiologyPerSignature));
```
Comment (kduyvesteyn) [inline orange/src/test/java/com/hartwig/hmftools/orange/algo/sigs/SigsInterpreterTest.java:29]: You could potentially expand on this test? E.g. assert that you do end up with APOBEC as signature for example?

Also, you could test the scenario where `null` is fed into the interpreter?

**Review by kduyvesteyn — APPROVED** [review]
Looks good!

Have a bunch of smaller comments, feel free to followup but also fine if you ignore!
